File: ch6/sa_lcp.java
// strncmp-in-java-SA-code-not-needed,java-LRS-minor-bug

import java.util.*;

class sa_lcp {
  Scanner scan;
  char T[];                        // the input string, up to 100K characters
  int n;                                             // the length of input string

  int[] RA, tempRA;             // rank array and temporary rank array
  Integer[] SA, tempSA;         // suffix array and temporary suffix array
  int[] c;                                         // for counting/radix sort

  char P[];     // the pattern string (for string matching)
  int m;             // the length of pattern string

  int[] Phi;    // for computing longest common prefix
  int[] PLCP;
  int[] LCP;    // LCP[i] stores the LCP between previous suffix "T + SA[i-1]" and current suffix "T + SA[i]"

  void countingSort(int k) {
    int i, sum, maxi = Math.max(300, n);   // up to 255 ASCII chars or length of n
    for (i = 0; i < 100010; i++) c[i] = 0;                // clear frequency table
    for (i = 0; i < n; i++)                    // count the frequency of each rank
      c[i + k < n ? RA[i + k] : 0]++;
    for (i = sum = 0; i < maxi; i++) {
      int t = c[i]; c[i] = sum; sum += t;
    }
    for (i = 0; i < n; i++)               // shuffle the suffix array if necessary
      tempSA[c[SA[i] + k < n ? RA[SA[i] + k] : 0]++] = SA[i];
    for (i = 0; i < n; i++)                          // update the suffix array SA
      SA[i] = tempSA[i];
  }

  void constructSA() {              // this version can go up to 100000 characters
    int i, k, r;
    for (i = 0; i < n; i++) RA[i] = T[i];                      // initial rankings
    for (i = 0; i < n; i++) SA[i] = i;          // initial SA: {0, 1, 2, ..., n-1}
    for (k = 1; k < n; k <<= 1) {            // repeat sorting process log n times
      countingSort(k);       // actually radix sort: sort based on the second item
      countingSort(0);               // then (stable) sort based on the first item
      tempRA[SA[0]] = r = 0;                  // re-ranking; start from rank r = 0
      for (i = 1; i < n; i++)                         // compare adjacent suffices
        tempRA[SA[i]] =      // if same pair => same rank r; otherwise, increase r
          (RA[SA[i]] == RA[SA[i-1]] && RA[SA[i]+k] == RA[SA[i-1]+k]) ? r : ++r;
      for (i = 0; i < n; i++)                          // update the rank array RA
        RA[i] = tempRA[i];
  } }

  void computeLCP() {
    int i, L;
    Phi[SA[0]] = -1;                                              // default value
    for (i = 1; i < n; i++)                                 // compute Phi in O(n)
      Phi[SA[i]] = SA[i-1];         // remember which suffix is behind this suffix
    for (i = L = 0; i < n; i++) {                  // compute Permuted LCP in O(n)
      if (Phi[i] == -1) { PLCP[i] = 0; continue; }                 // special case
      while (i + L < T.length && Phi[i] + L < T.length && T[i + L] == T[Phi[i] + L]) L++; // L will be increased max n times
      PLCP[i] = L;
      L = Math.max(L-1, 0);                          // L will be decreased max n times
    }
    for (i = 1; i < n; i++)                                 // compute LCP in O(n)
      LCP[i] = PLCP[SA[i]];   // put the permuted LCP back to the correct position
  }

  int strncmp(char[] a, int i, char[] b, int j, int n){
    for (int k=0; i+k < a.length && j+k < b.length; k++){
      if (a[i+k] != b[j+k]) return a[i+k] - b[j+k];
    }
    return 0;
  }

  int[] stringMatching() {                           // string matching in O(m log n)
    int lo = 0, hi = n-1, mid = lo;                 // valid matching = [0 .. n-1]
    while (lo < hi) {                                          // find lower bound
      mid = (lo + hi) / 2;                                   // this is round down
      int res = strncmp(T, SA[mid], P, 0, m);       // try to find P in suffix 'mid'
      if (res >= 0) hi = mid;             // prune upper half (notice the >= sign)
      else          lo = mid + 1;                // prune lower half including mid
    }                                           // observe `=' in "res >= 0" above
    if (strncmp(T,SA[lo], P,0, m) != 0) return new int[]{-1, -1};         // if not found
    int[] ans = new int[]{ lo, 0} ;

    lo = 0; hi = n - 1; mid = lo;
    while (lo < hi) {                 // if lower bound is found, find upper bound
      mid = (lo + hi) / 2;
      int res = strncmp(T, SA[mid], P,0, m);
      if (res > 0) hi = mid;                                   // prune upper half
      else         lo = mid + 1;                 // prune lower half including mid
    }                                // (notice the selected branch when res == 0)
    if (strncmp(T, SA[hi], P,0, m) != 0) hi--;                      // special case
    ans[1] = hi;
    return ans;
  } // return lower/upper bound as the first/second item of the pair, respectively

  void LRS() {                          // print out the length and the actual LRS
    int i, idx = 0, maxLCP = 0;

    for (i = 1; i < n; i++)                                                // O(n)
      if (LCP[i] > maxLCP) {
        maxLCP = LCP[i];
        idx = i;
      }

    System.out.printf("\nThe LRS is '%s' with length = %d\n\n",
      new String(T).substring(SA[idx], SA[idx]+maxLCP), maxLCP);
  }

  int owner(int idx) { return (idx < n-m-1) ? 1 : 2; }

  void LCS() {                          // print out the length and the actual LCS
    int i, j, maxLCP = 0, idx = 0;
    // not used in Java version
    // char ans[MAX_N];
    // strcpy(ans, "");
    
    //System.out.printf("\nRemember, T = '%s'\nNow, enter another string P:\n", new String(T));
    // T already has '.' at the back
    P = new String("CATA").toCharArray();
    m = P.length;
    T = (new String(T) + new String(P) + "#").toCharArray();   // append P and '#'
    n = T.length;                                                      // update n
    constructSA();                                                   // O(n log n)
    computeLCP();                                                          // O(n)
    System.out.printf("\nThe LCP information of 'T+P' = '%s':\n", new String(T));
    System.out.printf("i\tSA[i]\tLCP[i]\tOwner\tSuffix\n");
    for (i = 0; i < n; i++)
      System.out.printf("%2d\t%2d\t%2d\t%2d\t%s\n", i, SA[i], LCP[i], owner(SA[i]),
        new String(T, SA[i], T.length - SA[i]));

    for (i = 1, maxLCP = -1; i < n; i++)
      if (LCP[i] > maxLCP && owner(SA[i]) != owner(SA[i-1])) {  // different owner
        maxLCP = LCP[i];
        idx = i;
        // not used in Java version
        // strncpy(ans, T + SA[i], maxLCP);
        // ans[maxLCP] = 0;
      }

    System.out.printf("\nThe LCS is '%s' with length = %d\n",
      new String(T).substring(SA[idx], SA[idx] + maxLCP),
      maxLCP);
  }

  void run() {
    int MAX_N = 100010;
    c = new int[MAX_N];
    RA = new int[MAX_N];
    tempRA = new int[MAX_N];
    SA = new Integer[MAX_N];
    tempSA = new Integer[MAX_N];
    Phi = new int[MAX_N];
    PLCP = new int[MAX_N];
    LCP = new int[MAX_N];

    //System.out.printf("Enter a string T below, we will compute its Suffix Array:\n");
    T = new String("COMPETITIVE$").toCharArray();
    n = T.length;

    constructSA();                                                   // O(n log n)
    System.out.printf("The Suffix Array of string T = '%s' is shown below (O(n log n) version):\n", new String(T));
    System.out.printf("i\tSA[i]\tSuffix\n");
    for (int i = 0; i < n; i++)
     System.out.printf("%2d\t%2d\t%s\n", i, SA[i], new String(T, SA[i], T.length - SA[i]));

    computeLCP();                                                          // O(n)

    // LRS demo
    LRS();        // find the longest repeated substring of the first input string

    // stringMatching demo
    //System.out.printf("\nNow, enter a string P below, we will try to find P in T:\n");
    P = new String("A").toCharArray();
    int[] pos = stringMatching();
    if (pos[0] != -1 && pos[1] != -1) {
      System.out.printf("%s is found SA [%d .. %d] of %s\n", 
        new String(P), pos[0], pos[1], new String(T));
      System.out.printf("They are:\n");
      for (int i = pos[0]; i <= pos[1]; i++)
        System.out.printf("  %s\n", new String(T, SA[i], T.length - SA[i]));
    } else System.out.printf("%s is not found in %s\n", new String(P), new String(T));

    // LCS demo
    LCS();                    // find the longest common substring between T and P

    // note that the LRS and LCS demo are slightly different in Java version
  }

  public static void main(String[] args){
    new sa_lcp().run();
  }
}

File: ch6/sa_lcp.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<int, int> ii;
typedef vector<int> vi;

class SuffixArray {
private:
  vi RA;                                         // rank array

  void countingSort(int k) {                     // O(n)
    int maxi = max(300, n);                      // up to 255 ASCII chars
    vi c(maxi, 0);                               // clear frequency table
    for (int i = 0; i < n; ++i)                  // count the frequency
      ++c[i+k < n ? RA[i+k] : 0];                // of each integer rank
    for (int i = 0, sum = 0; i < maxi; ++i) {
      int t = c[i]; c[i] = sum; sum += t;
    }
    vi tempSA(n);
    for (int i = 0; i < n; ++i)                  // sort SA
      tempSA[c[SA[i]+k < n ? RA[SA[i]+k] : 0]++] = SA[i];
    swap(SA, tempSA);                            // update SA
  }

  void constructSA() {                           // can go up to 400K chars
    SA.resize(n);
    iota(SA.begin(), SA.end(), 0);               // the initial SA
    RA.resize(n);
    for (int i = 0; i < n; ++i) RA[i] = T[i];    // initial rankings
    for (int k = 1; k < n; k <<= 1) {            // repeat log_2 n times
      // this is actually radix sort
      countingSort(k);                           // sort by 2nd item
      countingSort(0);                           // stable-sort by 1st item
      vi tempRA(n);
      int r = 0;
      tempRA[SA[0]] = r;                         // re-ranking process
      for (int i = 1; i < n; ++i)                // compare adj suffixes
        tempRA[SA[i]] = // same pair => same rank r; otherwise, increase r
          ((RA[SA[i]] == RA[SA[i-1]]) && (RA[SA[i]+k] == RA[SA[i-1]+k])) ?
            r : ++r;
      swap(RA, tempRA);                          // update RA
      if (RA[SA[n-1]] == n-1) break;             // nice optimization
    }
  }

  void computeLCP() {
    vi Phi(n);
    vi PLCP(n);
    PLCP.resize(n);
    Phi[SA[0]] = -1;                             // default value
    for (int i = 1; i < n; ++i)                  // compute Phi in O(n)
      Phi[SA[i]] = SA[i-1];                      // remember prev suffix
    for (int i = 0, L = 0; i < n; ++i) {         // compute PLCP in O(n)
      if (Phi[i] == -1) { PLCP[i] = 0; continue; } // special case
      while ((i+L < n) && (Phi[i]+L < n) && (T[i+L] == T[Phi[i]+L]))
        ++L;                                     // L incr max n times
      PLCP[i] = L;
      L = max(L-1, 0);                           // L dec max n times
    }
    LCP.resize(n);
    for (int i = 0; i < n; ++i)                  // compute LCP in O(n)
      LCP[i] = PLCP[SA[i]];                      // restore PLCP
  }

public:
  const char* T;                                 // the input string
  const int n;                                   // the length of T
  vi SA;                                         // Suffix Array
  vi LCP;                                        // of adj sorted suffixes

  SuffixArray(const char* initialT, const int _n) : T(initialT), n(_n) {
    constructSA();                               // O(n log n)
    computeLCP();                                // O(n)
  }

  ii stringMatching(const char *P) {             // in O(m log n)
    int m = (int)strlen(P);                      // usually, m < n
    int lo = 0, hi = n-1;                        // range = [0..n-1]
    while (lo < hi) {                            // find lower bound
      int mid = (lo+hi) / 2;                     // this is round down
      int res = strncmp(T+SA[mid], P, m);        // P in suffix SA[mid]?
      (res >= 0) ? hi = mid : lo = mid+1;        // notice the >= sign
    }
    if (strncmp(T+SA[lo], P, m) != 0) return {-1, -1}; // if not found
    ii ans; ans.first = lo;
    hi = n-1;                                    // range = [lo..n-1]
    while (lo < hi) {                            // now find upper bound
      int mid = (lo+hi) / 2;
      int res = strncmp(T+SA[mid], P, m);
      (res > 0) ? hi = mid : lo = mid+1;         // notice the > sign
    }
    if (strncmp(T+SA[hi], P, m) != 0) --hi;      // special case
    ans.second = hi;
    return ans;                                  // returns (lb, ub)
  }                                              // where P is found

  ii LRS() {                                     // (LRS length, index)
    int idx = 0, maxLCP = -1;
    for (int i = 1; i < n; ++i)                  // O(n), start from i = 1
      if (LCP[i] > maxLCP)
        maxLCP = LCP[i], idx = i;
    return {maxLCP, idx};
  }

  ii LCS(int split_idx) {                        // (LCS length, index)
    int idx = 0, maxLCP = -1;
    for (int i = 1; i < n; ++i) {                // O(n), start from i = 1
      // if suffix SA[i] and suffix SA[i-1] came from the same string, skip
      if ((SA[i] < split_idx) == (SA[i-1] < split_idx)) continue;
      if (LCP[i] > maxLCP)
        maxLCP = LCP[i], idx = i;
    }
    return {maxLCP, idx};
  }
};

const int MAX_N = 450010;                        // can go up to 450K chars

char T[MAX_N];
char P[MAX_N];
char LRS_ans[MAX_N];
char LCS_ans[MAX_N];

int main() {
  freopen("sa_lcp_in.txt", "r", stdin);
  scanf("%s", &T);                               // read T
  int n = (int)strlen(T);                        // count n
  T[n++] = '$';                                  // add terminating symbol
  SuffixArray S(T, n);                           // construct SA+LCP

  printf("T = '%s'\n", T);
  printf(" i SA[i] LCP[i]   Suffix SA[i]\n");
  for (int i = 0; i < n; ++i)
    printf("%2d    %2d    %2d    %s\n", i, S.SA[i], S.LCP[i], T+S.SA[i]);

  // String Matching demo, we will try to find P in T
  strcpy(P, "A");
  auto [lb, ub] = S.stringMatching(P);
  if ((lb != -1) && (ub != -1)) {
    printf("P = '%s' is found SA[%d..%d] of T = '%s'\n", P, lb, ub, T);
    printf("They are:\n");
    for (int i = lb; i <= ub; ++i)
      printf("  %s\n", T+S.SA[i]);
  }
  else
    printf("P = '%s' is not found in T = '%s'\n", P, T);

  // LRS demo, find the LRS of T
  auto [LRS_len, LRS_idx] = S.LRS();
  strncpy(LRS_ans, T+S.SA[LRS_idx], LRS_len);
  printf("The LRS is '%s' with length = %d\n", LRS_ans, LRS_len);

  // LCS demo, find the LCS of (T, P)
  strcpy(P, "CATA");
  int m = (int)strlen(P);
  strcat(T, P);                                  // append P to T
  strcat(T, "#");                                // add '#' at the back
  n = (int)strlen(T);                            // update n

  // reconstruct SA of the combined strings
  SuffixArray S2(T, n);                          // reconstruct SA+LCP
  int split_idx = n-m-1;
  printf("T+P = '%s'\n", T);
  printf(" i SA[i] LCP[i] From  Suffix SA[i]\n");
  for (int i = 0; i < n; ++i)
    printf("%2d    %2d    %2d    %2d    %s\n", 
      i, S2.SA[i], S2.LCP[i], S2.SA[i] < split_idx ? 1 : 2, T+S2.SA[i]);

  auto [LCS_len, LCS_idx] = S2.LCS(split_idx);
  strncpy(LCS_ans, T+S2.SA[LCS_idx], LCS_len);
  printf("The LCS is '%s' with length = %d\n", LCS_ans, LCS_len);
  
  return 0;
}
